firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js

```diff
@@ -578,6 +578,70 @@ window.__firefox__.includeOnce("LoginsHelper", function() {
     },
   }
 
+  // define the field types for focus events
+  const FocusFieldType = {
+    username: "username",
+    password: "password"
+  };
+
+  // Extended the LoginManagerContent object with a new method for focus events
+  LoginManagerContent._addFocusListenerBasedOnHeuristics = function (form) {
+    // common form structure
+    var commonStructures = `input[type="text"], input[type="email"], input:not([type])`
+    var possibleUsernameFields = Array.from(form.querySelectorAll(commonStructures));
+    var passwordFields = form.querySelectorAll('input[type="password"]');
+
+    // add listener to password fields
+    passwordFields.forEach((field) => this._addFocusListener(field, FocusFieldType.password));
+
+    // determine possible username field
+    possibleUsernameFields = possibleUsernameFields.filter(field =>
+      field.placeholder.toLowerCase().includes("user") ||
+      field.placeholder.toLowerCase().includes("email") ||
+      field.id.toLowerCase().includes("user") ||
+      field.id.toLowerCase().includes("email") ||
+      field.name.toLowerCase().includes("user") ||
+      field.name.toLowerCase().includes("email"));
+
+    // prioritize fields closer to the first password field, if available
+    if (passwordFields.length > 0) {
+      const firstPasswordIndex = Array.from(form.elements).indexOf(passwordFields[0]);
+      possibleUsernameFields = possibleUsernameFields.filter(field =>
+        Array.from(form.elements).indexOf(field) < firstPasswordIndex);
+    }
+
+    // attach listeners to possible username fields, if not already focused on password fields
+    possibleUsernameFields.forEach((field) => this._addFocusListener(field, FocusFieldType.username));
+
+    // no clear username field is identified & password fields are present,
+    // attach to the first text input before the password field as a fallback
+    if (possibleUsernameFields.length === 0 && passwordFields.length > 0) {
+      let precedingTextFields = Array.from(passwordFields[0].parentNode.querySelectorAll('input[type="text"], input[type="email"], input:not([type])'));
+      if (precedingTextFields.length > 0) {
+        // Last text input before the first password field
+        let assumedUsernameField = precedingTextFields[precedingTextFields.length - 1];
+        this._addFocusListener(assumedUsernameField, FocusFieldType.username);
+      }
+    }
+  };
+
+  LoginManagerContent._addFocusListener = function (field, fieldType) {
+    field.addEventListener("focus", function () {
+      // send message with field focus information
+      var messageData = {
+        type: "fieldFocus",
+        fieldType: fieldType
+      };
+      webkit.messageHandlers.loginsManagerMessageHandler.postMessage(messageData);
+    });
+  };
+
+  var originalFindLogins = findLogins;
+  findLogins = function (form) {
+    originalFindLogins(form);
+    LoginManagerContent._addFocusListenerBasedOnHeuristics(form); // Use the new method here
+  };
+    
```

Comment:
See my comment above. Attaching focus event listeners on each field might be a bit overkill. Let's flip the logic here. Instead of trying to find the inputs and attaching a `focus` event handler, let's have a `focusin` event on the document and figure out later what was focused ( using `focusin` instead of `focus` because `focus` doesn't bubble )